src/Raven.Server/Documents/Archive/ArchivalStorage.cs

```diff
@@ -0,0 +1,356 @@
+            var clonedCountersGroupsValues = new List<BlittableJsonReaderObject>(counterGroups.Count);
+            foreach (var counterGroup in counterGroups)
+                clonedCountersGroupsValues.Add(counterGroup.Values.Clone(context));
+                                            
+            var mergedCounterGroupsValues = BuildMergedCounterGroupValues(context, clonedCountersGroupsValues, documentId);
+                                                
+            _database.DocumentsStorage.CountersStorage.DeleteCountersForDocument(context, documentId, new CollectionName(collectionName));
+                                            
+            var newCounterEtag = _database.DocumentsStorage.GenerateNextEtag();
+            var newCounterChangeVector = _database.DocumentsStorage.GetNewChangeVector(context, newCounterEtag);
+            _database.DocumentsStorage.CountersStorage.PutCounters(context, documentId, archivedCollectionName, newCounterChangeVector , mergedCounterGroupsValues, newCounterEtag);
+        }
+
+        private void ArchiveDocumentTimeSeriesIfNeeded(DocumentsOperationContext context, Document doc, string documentId, string collectionName, string archivedCollectionName)
+        {
+            
+        }
+        
+        private void ArchiveDocumentRevisionsIfNeeded(DocumentsOperationContext context, Document doc, string documentId, string collectionName, string archivedCollectionName)
+        {
+            
+        }
+        
+        private void ArchiveDocumentAttachmentsIfNeeded(DocumentsOperationContext context, Document doc, string documentId, string collectionName, string archivedCollectionName)
+        {
+                    
+        }
+
+        private bool MigrateDocument(DocumentsOperationContext context, (Slice LowerId, string Id) ids, DateTime currentTime)
+        {
+            if (ids.Id == null)
+                throw new InvalidOperationException($"Couldn't archive the document. Document id is null. Lower id is {ids.LowerId}");
+
+            using (var doc = _database.DocumentsStorage.Get(context, ids.LowerId, DocumentFields.Data, throwOnConflict: true))
+            {
+                if (doc == null || !doc.TryGetMetadata(out var metadata))
+                {
+                    throw new InvalidOperationException($"Failed to fetch the metadata of document '{ids.Id}'");
+                }
+                if (HasPassed(metadata, currentTime) == false) return false;
+                        
+                var collectionName = metadata[Constants.Documents.Metadata.Collection].ToString();
+                string archivedCollectionName = $"{collectionName}Archived";
+
+                // Modify
+                metadata.Modifications = new DynamicJsonValue(metadata);
+                metadata.Modifications[Constants.Documents.Metadata.Collection] = archivedCollectionName;
+                metadata.Modifications.Remove(Constants.Documents.Metadata.Archive);
```

Comment:
We need a marker in metadata saying this is archived?
Should we also add the original collection here?

Reply:
That'd be a very informative entry in the metadata, but may be redundant.
We already had named the collection with our special suffix.
The one case that comes to my mind quickly -> the user has defined a collection, which name contains the archived docs suffix, but it wasn't yet archived indeed, of course.

On the other hand, I hadn't to re-check if the doc has been already archived.
Its retirement timestamp is deleted on the process, so the user would need to intentionally do that again on archived collection docs. But is there any sense to do that?

Reply:
It could be helpful while debugging - if we're doing wrong (archiving archived docs over), it'd inform us about unexpected behavior.

Reply:
Combine that with the custom postfix, and you can't *rely* on that.
I think that it would also be nice to *know* that a document is archived, so we can tell "are you sure you want to modify it?"

Reply:
Fixed in https://github.com/ravendb/ravendb/pull/15715/commits/eaee150b5fc61479d35bdc0113c70a5d73dfcb5d